From 6fb2149fc5c7b1805c138a52990305a4f53170fa Mon Sep 17 00:00:00 2001
From: Hongxu Jia <hongxu.jia@eng.windriver.com>
Date: Thu, 11 Dec 2025 18:38:24 -0800
Subject: [PATCH] meson: Add gir_dir_prefix option to override gir default
 location

Originally, the .gir files would always get installed into the default
location (datadir).

This commit adds meson option (gir_dir_prefix) to override it as intermediate
prefix for gir installation under ${prefix}, such as

  $ meson setup -Dgir_dir_prefix=lib64

The .gir files are installed to /usr/lib64/gir-1.0

  Installing gst/Gst-1.0.gir to /usr/lib64/gir-1.0

Upstream-Status: Submitted [https://gitlab.freedesktop.org/gstreamer/gstreamer/-/merge_requests/10363]

Signed-off-by: Hongxu Jia <hongxu.jia@eng.windriver.com>
---
 gst/meson.build                 |  1 +
 libs/gst/base/meson.build       |  1 +
 libs/gst/check/meson.build      |  1 +
 libs/gst/controller/meson.build |  1 +
 libs/gst/net/meson.build        |  1 +
 meson.build                     | 23 ++++++++++++++++++++++-
 meson_options.txt               |  3 +++
 7 files changed, 30 insertions(+), 1 deletion(-)

diff --git a/gst/meson.build b/gst/meson.build
index c69a6dc..ebe673d 100644
--- a/gst/meson.build
+++ b/gst/meson.build
@@ -322,6 +322,7 @@ if build_gir
     'export_packages' : pkg_name,
     'includes' : ['GLib-2.0', 'GObject-2.0', 'GModule-2.0' ],
     'install' : true,
+    'install_dir_gir': gst_girdir,
     'extra_args' : gst_gir_extra_args,
   }
   library_def += {'gir': [gir]}
diff --git a/libs/gst/base/meson.build b/libs/gst/base/meson.build
index b7aca73..2fb0590 100644
--- a/libs/gst/base/meson.build
+++ b/libs/gst/base/meson.build
@@ -89,6 +89,7 @@ if build_gir
     'include_directories' : [configinc, libsinc, privinc],
     'includes' : ['GLib-2.0', 'GObject-2.0', 'GModule-2.0', 'Gst-1.0'],
     'install' : true,
+    'install_dir_gir': gst_girdir,
     'extra_args' : gst_gir_extra_args,
   }
 
diff --git a/libs/gst/check/meson.build b/libs/gst/check/meson.build
index 679a15d..e934526 100644
--- a/libs/gst/check/meson.build
+++ b/libs/gst/check/meson.build
@@ -86,6 +86,7 @@ if build_gir
     'include_directories' : [configinc, libsinc, privinc],
     'includes' : ['GLib-2.0', 'GObject-2.0', 'GModule-2.0', 'Gst-1.0'],
     'install' : true,
+    'install_dir_gir': gst_girdir,
     'extra_args' : gst_gir_extra_args,
   }
   library_def += {'gir': [gir]}
diff --git a/libs/gst/controller/meson.build b/libs/gst/controller/meson.build
index 9d38af8..411b951 100644
--- a/libs/gst/controller/meson.build
+++ b/libs/gst/controller/meson.build
@@ -81,6 +81,7 @@ if build_gir
     'include_directories' : [configinc, libsinc, privinc],
     'includes' : ['GLib-2.0', 'GObject-2.0', 'GModule-2.0', 'Gst-1.0'],
     'install' : true,
+    'install_dir_gir': gst_girdir,
     'extra_args' : gst_gir_extra_args,
   }
   library_def += {'gir': [gir]}
diff --git a/libs/gst/net/meson.build b/libs/gst/net/meson.build
index 60ed92c..b81139e 100644
--- a/libs/gst/net/meson.build
+++ b/libs/gst/net/meson.build
@@ -67,6 +67,7 @@ if build_gir
     'include_directories' : [configinc, libsinc],
     'includes' : ['GLib-2.0', 'GObject-2.0', 'GModule-2.0', 'Gio-2.0', 'Gst-1.0'],
     'install' : true,
+    'install_dir_gir': gst_girdir,
     'extra_args' : gst_gir_extra_args,
   }
   library_def += {'gir': [gir]}
diff --git a/meson.build b/meson.build
index e2127f6..7ecfc2d 100644
--- a/meson.build
+++ b/meson.build
@@ -629,6 +629,27 @@ if static_build
   gst_c_args += ['-DGST_STATIC_COMPILATION']
 endif
 
+gir_dir_prefix = get_option('gir_dir_prefix')
+if gir_dir_prefix == '' or gir_dir_prefix == get_option('datadir')
+  gir_dir_prefix = get_option('datadir')
+  gir_dir_pc_prefix = '${datadir}'
+else
+  gir_dir_pc_prefix = '${prefix}' / gir_dir_prefix
+endif
+
+# This is effectively the GIR XML format major version
+gir_suffix = 'gir-1.0'
+
+gst_girdir = prefix / gir_dir_prefix / gir_suffix
+
+
+if build_gir
+  gst_c_args += [
+    '-DGIR_SUFFIX="@0@"'.format(gir_suffix),
+    '-DGIR_DIR="@0@"'.format(gst_girdir),
+  ]
+endif
+
 # Used in gst/parse/meson.build and below
 python3 = import('python').find_installation()
 
@@ -680,7 +701,7 @@ pkgconfig_variables = [
   'exec_prefix=${prefix}',
   'toolsdir=${exec_prefix}/bin',
   'pluginsdir=${libdir}/gstreamer-1.0',
-  'girdir=${datadir}/gir-1.0',
+  'girdir=' + gir_dir_pc_prefix / gir_suffix,
   'typelibdir=${libdir}/girepository-1.0',
   'pluginscannerdir=${libexecdir}/gstreamer-1.0'
 ]
diff --git a/meson_options.txt b/meson_options.txt
index 78af552..97c03c6 100644
--- a/meson_options.txt
+++ b/meson_options.txt
@@ -32,6 +32,9 @@ option('examples', type : 'feature', value : 'auto', yield : true)
 option('tests', type : 'feature', value : 'auto', yield : true)
 option('benchmarks', type : 'feature', value : 'auto', yield : true)
 option('tools', type : 'feature', value : 'auto', yield : true)
+option('gir_dir_prefix',
+       type: 'string',
+       description: 'Intermediate prefix for gir installation under ${prefix}')
 option('introspection', type : 'feature', value : 'auto', yield : true, description : 'Generate gobject-introspection bindings')
 option('nls', type : 'feature', value : 'auto', yield: true, description : 'Enable native language support (translations)')
 option('extra-checks', type : 'feature', value : 'enabled', yield : true, description : 'Enable extra runtime checks')
-- 
2.34.1

